### PR TITLE
Assembly crate module reorganization

### PR DESCRIPTION
**Pull request by bobbinth**
This PR re-organized modules in the `assembly` crate and cleans up the crate's exports. Specifically, I introduced `ast` module and moved what was under `parsers` under this module (and re-shuffled modules a bit to make the organization a bit more logical).

I also removed duplicated exports. The create should still export all the same structures/enums/traits as before, but depending on how downstream creates imported these components, this PR may contain breaking changes.

### CODE REVIEW

File: assembly/src/assembler/mod.rs [hunk 1]
```diff
@@ -28,9 +28,9 @@ mod tests;
 
 // ASSEMBLER
 // ================================================================================================
-/// Miden Assembler which can be used to convert Miden assembly source code into program MAST (
-/// represented by the [Program] struct). The assembler can be instantiated in several ways using
-/// a "builder" patter. Specifically:
+/// Miden Assembler which can be used to convert Miden assembly source code into program MAST.
+///
+/// The assembler can be instantiated in several ways using a "builder" patter. Specifically:
```
Comment (frisitano) [inline assembly/src/assembler/mod.rs:33]: typo: patter -> pattern

File: assembly/src/ast/code_body.rs [hunk 2]
```diff
@@ -60,31 +32,41 @@ impl CodeBody {
...
         self
     }
 
     // STATE MUTATORS
     // --------------------------------------------------------------------------------------------
 
-    /// Pushes the provided location to the structure.
+    /// Adds the provided location to the end of location list.
+    ///
+    /// It is expected that prior to calling this method number of nodes and locations contained
```
Comment (frisitano) [inline assembly/src/ast/code_body.rs:55]: Should this be "... prior to calling this method the number of nodes ...".

File: assembly/src/errors.rs [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
+    crypto::hash::RpoDigest, tokens::SourceLocation, LibraryNamespace, ProcedureId, String,
+    ToString, Token, Vec,
 };
 use core::fmt;
+use vm_core::utils::write_hex_bytes;
 
 // ASSEMBLY ERROR
 // ================================================================================================
 
+/// An error which can be generating while compiling a Miden assembly program into MAST.
```
Comment (frisitano) [inline assembly/src/errors.rs:11]: typo: generating -> generated

File: assembly/src/errors.rs [hunk 4]
```diff
@@ -156,6 +158,7 @@ impl std::error::Error for AssemblyError {}
 // PARSING ERROR
 // ================================================================================================
 
+/// An error which can be generating while parsing a Miden assembly source code into an AST.
```
Comment (frisitano) [inline assembly/src/errors.rs:161]: typo: generating -> generated

File: assembly/src/ast/nodes/serde/mod.rs [hunk 5]
```diff
@@ -1,10 +1,9 @@
-use super::{
-    ByteReader, ByteWriter, CodeBody, Deserializable, DeserializationError, Felt, Instruction,
-    Node, ProcedureId, RpoDigest, Serializable,
-};
+use super::{CodeBody, Felt, Instruction, Node, ProcedureId, RpoDigest};
 use crate::MAX_PUSH_INPUTS;
 use num_enum::TryFromPrimitive;
 
+use vm_core::utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable};
```
Comment (frisitano) [inline assembly/src/ast/nodes/serde/mod.rs:5]: Maybe we should consider standardising the grouping of imports across our crates.  I believe `rustfmt` has an option for this (but I don't think it's stable) - https://rust-lang.github.io/rustfmt/?version=v1.4.38&search=#group_imports.

**Review by frisitano — APPROVED** [review]
Looks good! Left a few comments inline relating to typos.

**Review by jjcnn — APPROVED** [review]
LGTM. This breaks a few things in the PR I'm working on, so I'd like this to be merged asap.

File: assembly/src/ast/nodes/serde/mod.rs (hunk 5 above)
Comment (bobbinth) [reply assembly/src/ast/nodes/serde/mod.rs:5]: I tried this at some point and it looked a bit "messy" to me. But not a strong preference and if others like it, I'm happy to revisit.

**Comment by bobbinth** [comment]
> This breaks a few things in the PR I'm working on, so I'd like this to be merged asap.

Merged!

**Pull request merged**